### PR TITLE
ExecutionProvider API refactor - replace OrtMemoryInfo with OrtDevice

### PR DESCRIPTION
**Pull request by jslhcl**
### Description
ExecutionProvider API refactor - replace OrtMemoryInfo with OrtDevice

### Motivation and Context
Currently “Location” is represented as ORTMemoryInfo, which is OrtDevice + OrtMemType, while OrtDevice is represent as DeviceType + DeviceId + MemType. As we can see there is some unnecessary hierarchy, the proposal is to make it a clear definition that to use OrtDevice as an abstraction for Location

### CODE REVIEW

**Comment by jslhcl** [comment]
>   static const DeviceType FPGA = 2;

No FPGA device in EP?

---
Refers to: include/onnxruntime/core/framework/ortdevice.h:18 in 9bd7286. [](commit_id = 9bd72862610e301898ce332ca2bef16268976c96, deletion_comment = False)

File: include/onnxruntime/core/framework/execution_provider.h [hunk 1]
```diff
@@ -60,11 +60,15 @@ enum class DataLayout {
...
-      : type_{type} {
+      : IExecutionProvider(type, OrtDevice(), use_metadef_id_creator) {}
+
+  IExecutionProvider(const std::string& type, OrtDevice device, bool use_metadef_id_creator = false)
+      : default_device_(device), type_{type} {
     if (use_metadef_id_creator) {
       metadef_id_generator_ = std::make_unique<ModelMetadefIdGenerator>();
     }
   }
+  OrtDevice default_device_;
```
Comment (souptc) [inline include/onnxruntime/core/framework/execution_provider.h:71]: this should be private?

and please add comments for the new member/method.

Comment (jslhcl) [reply include/onnxruntime/core/framework/execution_provider.h:75]: it is protected. 

Comments added

**Review by souptc — DISMISSED** [review]
:shipit:

**Comment by souptc** [comment]
>   static const DeviceType FPGA = 2;

FPGA is resvered for some 1p hardware.

---
In reply to: [1517106603](https://github.com/microsoft/onnxruntime/pull/15618#issuecomment-1517106603) [](http://example.com/codeflow?ancestors=1517106603,1517106603)

---
Refers to: include/onnxruntime/core/framework/ortdevice.h:18 in 9bd7286. [](commit_id = 9bd72862610e301898ce332ca2bef16268976c96, deletion_comment = False)

**Review by souptc — APPROVED** [review]
:shipit:

**Pull request merged**